### DISCUSSION TITLE
MET error when loading netCDF file

### DISCUSSION THREAD

**Discussion by bmatilla** [dtcenter/METplus#2274] [File I/O] [answered]
Hi there,

I am attempting to load in a netCDF file into plot_data_plane just to see if I can generate an output image. However, when I try to run plot_data_plane with that netCDF file, the following error is displayed:

ERROR  : MetNcFile::data(NcVar *, const LongArray &, DataPlane &) const -> star found in bad slot

It has been a rather long time since I’ve used MET (just now getting spun back up using MET 11.0.2 as part of the coordinated release for METplus 5.0) and am not too sure what this error means, so I was looking for more of an explanation to this and if you have any ideas on how I can get around this error. Any help is greatly appreciated and feel free to let me know of any questions you may have.

**Comment by JohnHalleyGotway** [comment]
Hi Brian, thanks for posting your question to GitHub discussions.

I searched the earlier discussions and found this one: https://github.com/dtcenter/METplus/discussions/1411

@j-opatz provided a great description of the problem and solution there. Please take a look and let me know if that answers it. If not, please post a sample data file to our anonymous ftp site following [these instructions](https://github.com/dtcenter/METplus/discussions/954), along with the command you're running to produce the error.

I do want to note that MET 11.0.0 includes [this enhancement](https://github.com/dtcenter/MET/issues/1815) which makes requesting data from NetCDF files much more intuitive.

If reading data from a NetCDF variable like this `float temperature(time, pressure, y, x) ;`, previously you'd have to specify integer **INDICES** for each of the dimensions, like the:
```
plot_data_plane sample.nc plot.ps 'name="temperature"; level="(0,3,*,*)";'
```
That'd extract the 2D field of data for the first time step and fourth pressure level (the integers are 0-based). After this change, you can specify their **VALUES** instead:
```
plot_data_plane sample.nc plot.ps 'name="temperature"; level="(@20230315_12,@850,*,*)";'
```
With this, MET will figure out which `time` and `pressure` levels to use based on the values you specify for each dimension.

**Comment by bmatilla** [comment]
Hi John,

Just got to seeing your response this afternoon... Thanks very much for getting back to me on this. Those instructions are very useful and I'll look to those as a starting point. I'll give that a spin with some of the data I've got here and follow up with you soon if I come across issues.

**Comment by bmatilla** [comment]
Hi again John,

So I've had a look at the response in the earlier discussion you mentioned and tried to get the file I was looking to get plot_data_plane running with. Unfortunately, I have not been successful in doing so. I've gone ahead and placed a copy of the file I have along with the command I am using to run plot_data_plane with in a separate text file through the ftp server. Please let me know if you have any issues seeing the files or any other questions that come up.

**Comment by JohnHalleyGotway** [comment, marked as answer]
@bmatilla sorry for the delay in responding. I was out of the office mid-week.

I see the source of the issue here. The MET software supports a few different flavors of gridded NetCDF files:
1. [CF-compliant](url) NetCDF files
2. The NetCDF output of the wrf_interp utility
3. MET's own internal NetCDF format

For (3), we'd actually like to stop using a custom internal format and switch entirely to follow the CF-convention. But that change hasn't bubbled up to the top of the priority list yet.

MET has separate library code for reading data from each of these flavors. The challenge is after opening up the NetCDF file, how to determine which library should be used. If you're curious about that logic, it lives in the [grd_file_type()](https://github.com/dtcenter/MET/blob/main_v11.1/src/libcode/vx_data2d_factory/data2d_factory_utils.cc#L57) utility function.

Your NetCDF file straddles the CF-convention and the internal MET format. Since the global attributes include `MET_version`, we assume that it follow the internal MET format.
```
// global attributes:
		:MET_version = "V11.0.2" ;
```
But you're specifying timing information in a CF-convention type of way. And that leads to the error.

You have 2 options to fix this:
1. Add `file_type = NETCDF_NCCF;` to your command line to explicitly tell MET to interpret this data as following the CF-convention:
```
plot_data_plane mrms_mergedazshear_0to2_20180502_161012.nc plot.ps 'name="MergedAzShear_0to2km"; level="(0,*,*)"; file_type=NETCDF_NCCF;'
```
![Screen Shot 2023-08-03 at 11 41 32 AM](https://github.com/dtcenter/METplus/assets/21087144/7811bcb6-304b-4332-adb2-044e2e59c661)
The downside is that you'd need to include that `file_type` option every time you read data from these files.
2. The better solution is to replace `MET_version = "V11.0.2" ;` with `Conventions = "CF-1.10";`. The actual CF version number doesn't matter all that much. 1.10 is the current version. Here's how you could do this using the `ncatted` utility and then plot the result *without* needing the `file_type` option:
```
ncatted -a MET_version,global,d,c,'' -a Conventions,global,a,c,'CF-1.10' \
   mrms_mergedazshear_0to2_20180502_161012.nc mrms_mergedazshear_0to2_20180502_161012_cf.nc
plot_data_plane mrms_mergedazshear_0to2_20180502_161012_cf.nc plot_cf.ps name="MergedAzShear_0to2km"; level="(0,*,*)"; 
```

The presence of the `Conventions = "CF-*` global attribute tells MET to interpret the file as being CF-compliant. So you would not need to specify the `file_type` option.

Hope that helps.

**Comment by bmatilla** [comment]
Hi John,

No worries; I appreciate your detailed response. It appears that did the trick on my end as I'm able to match your outcome! I'll look to modify my code so that I can rewrite the output files to include the updated CF convention attribute as in the latter of the two options. Thanks for also passing along the logic for the libraries. I'll need to take a more detailed look to learn more about that. For now, I'll mark your above response as the answer. Thanks very much once again!
